src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseReader.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Data.Common;
+
+namespace Microsoft.ML.Data
+{
+    public abstract class DatabaseReader : IDisposable
```

Comment:
I ended up going with this layout after experimentation with several different models.

This, I believe, should be a good compromise over tracking the least amount of state and requiring the end user to implement the least amount of code.